hack/collect-logs.sh

```diff
@@ -0,0 +1,106 @@
+  --all-namespaces -o yaml >> "${ARTIFACTS}/logs/capz.info" || true
+
+  # dump images info
+  {
+   echo "images in docker"
+   docker images
+   echo "images from bootstrap using containerd CLI"
+   docker exec kind-control-plane ctr -n k8s.io images list
+   echo "images in bootstrap cluster using kubectl CLI"
+   (kubectl get pods --all-namespaces -o json \
+   | jq --raw-output '.items[].spec.containers[].image' | sort)
+   echo "images in deployed cluster using kubectl CLI"
+   (kubectl --kubeconfig=$CLUSTER_NAME.kubeconfig get pods --all-namespaces -o json \
+   | jq --raw-output '.items[].spec.containers[].image' | sort)
+ } >> "${ARTIFACTS}/logs/images.info"
+  
+  # dump cluster info for kind
+  {
+    echo "kind cluster-info"
+    kubectl cluster-info dump
+  } >> "${ARTIFACTS}/logs/kind-cluster.info"
+
+  # dump cluster info for capz
+  {
+    echo "=== VMs in ${AZURE_RESOURCE_GROUP}  ==="
+    az vm list --resource-group "${AZURE_RESOURCE_GROUP}"
+    echo "=== cluster-info dump ==="
+    kubectl --kubeconfig=$CLUSTER_NAME.kubeconfig cluster-info dump
+  } >> "${ARTIFACTS}/logs/capz-cluster.info"
+
+  # export all logs from kind
+  kind "export" logs --name="kind" "${ARTIFACTS}/logs"
+
+  nodes=$(az vm list --resource-group ${AZURE_RESOURCE_GROUP} --query "[?tags.\"sigs.k8s.io_cluster-api-provider-azure_cluster_capi-quickstart\" == 'owned'].name" -o tsv)
+  declare -a nodeList=( $( echo $nodes | cut -d' ' -f1- ) )
+  # We used to pipe this output to 'tail -n +2' but for some reason this was sometimes (all the time?) only finding the
+  # bastion host. For now, omit the tail and gather logs for all VMs that have a private IP address. This will include
+  # the bastion, but that's better than not getting logs from all the VMs.
+  for node in "${nodeList[@]}"
+  do
+    echo "collecting logs from ${node}"
+    dir="${ARTIFACTS}/logs/${node}"
+    mkdir -p "${dir}"
+    ssh-to-node "${node}" "sudo journalctl --output=short-precise -k" > "${dir}/kern.log" || true
+    ssh-to-node "${node}" "sudo journalctl --output=short-precise" > "${dir}/systemd.log" || true
+    ssh-to-node "${node}" "sudo crictl version && sudo crictl info" > "${dir}/containerd.info" || true
+    ssh-to-node "${node}" "sudo journalctl --no-pager -u cloud-final" > "${dir}/cloud-final.log" || true
+    ssh-to-node "${node}" "sudo journalctl --no-pager -u kubelet.service" > "${dir}/kubelet.log" || true
+    ssh-to-node "${node}" "sudo journalctl --no-pager -u containerd.service" > "${dir}/containerd.log" || true
+  done
+}
+
+# SSH to a node by instance-id ($1) and run a command ($2).
+function ssh-to-node() {
+  local node="$1"
+  local cmd="$2"
+
+  #user=azureuser
+  #ssh_key_pem="/tmp/${AWS_SSH_KEY_NAME}.pem"
+  az vm run-command invoke -g ${AZURE_RESOURCE_GROUP} -n ${node} --command-id RunShellScript --scripts ${cmd}
```

Comment:
@devigned @CecileRobertMichon do either of you know of a better way for this? The output we get from this is not pretty.

Reply:
The output of `az vm run-command invoke` has a limit of 4096 characters, which isn't sufficient for most of the logs in our use case :/ Also, with the introduction of machinepool, we would have to use `az vmss run-command invoke` instead.

Another way we can collect logs is through a combination of DaemonSet and bash script as mentioned in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/534#issuecomment-619145321. The idea is to deploy a DaemonSet and use the pods on each node as a proxy to execute commands like `kubectl exec <daemonset pod> -- sudo journalctl ... > logfile` (see https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/hack/log-dump for more details).

This way, the output isn't limited to 4096 characters and we don't have to care about the VM type of the underlying nodes (VM / VMSS).

Reply:
That's a great idea Ernest. Plus hopefully it won't take as long as az vm run command